syntax = "proto3";

package combination.graph;

import "base.proto";
import "google/protobuf/empty.proto";
import "base_operate.proto";

option java_multiple_files = true;
option java_package = "com.github.churunfa.switchautoweb.combination.graph";
option java_outer_classname = "CombinationGraphProto";

service CombinationGraphService {
    rpc GetAllProject (google.protobuf.Empty) returns (GetAllProjectResponse) {}
    rpc GetAllGraph (StringValue) returns (GetAllGraphResponse) {}
    rpc GetGraphById (IntValue) returns (CombinationGraph) {}
    rpc InsertGraph (CombinationGraph) returns (base.SimpleResponse) {}
    rpc UpdateGraph (CombinationGraph) returns (base.SimpleResponse) {}
    rpc DeleteGraph (IntValue) returns (base.SimpleResponse) {}
    rpc ExecGraph(CombinationGraph) returns (base.SimpleResponse) {};
    rpc ExecGraphById(IntValue) returns (base.SimpleResponse) {};
}

message Combination {
    int32 id = 1;
    string project_name = 2;
    string combination_name = 3;
    string desc = 4;
    int32 min_time = 5;
}

message CombinationNode {
    int32 node_id = 1;
    base.operate.BaseOperate base_operate = 2;
    string params = 3;
    int32 exec_hold_time = 4;
    int32 reset_hold_time = 5;
    int32 loop_cnt = 6;
    bool auto_reset = 7;
}

message CombinationEdge {
    int32 edge_id = 1;
    int32 from_node_id = 2;
    int32 next_node_id = 3;

}

message CombinationGraph {
    Combination combination = 1;
    repeated CombinationNode combination_node = 2;
    repeated CombinationEdge combination_edges = 3;
}

message StringValue {
    string value = 1;
}

message IntValue {
    int32 value = 1;
}

message GetAllProjectResponse {
    repeated string project_names = 1;
}

message GetAllGraphResponse {
    repeated Combination combinations = 1;
}